### DISCUSSION TITLE
High memory usage when using distroless/static build

### DISCUSSION THREAD

**Discussion by rubenelshof** [jesec/flood#188] [Q&A] [answered]
Hi,

When I use the distroless docker image or download the static build in the latest release (4.4.0) and run flood I experience high memory usage.
Flood will go to about 4gb.
I dont experience this when using the release or rtorrent docker images.

Is this normal behaviour when going with distroless/static builds?

**Comment by jesec** [comment, marked as answer]
By default, like all language runtime with JIT capabilities, Node runtime tries to utilize memory available to it to optimize the performance of the application.

V8 engine reserves its own range of address spaces at launch. Such memory is neither allocated or committed. It will not block other programs from using the memory. 

You may check the real memory consumption with `sudo pmap -x <pid>`. It is also noticeable that the total memory available of your system will not decrease in that large amount.

**Reply by jesec** [reply]
It is regrettable that the default columns of `ps` can be confusing. Many Linux beginners got confused over `VSZ` and `RSS`. 

`VSZ` is NOT the real memory consumption. It is `RSS` that indicates the memory used by a certain process. 

Anyways, it should be easy, with just a little bit more attention, to notice `%MEM`, which indicates total memory used by a process in percentage. 

Apparently Windows Task Manger is much more user-friendly...

![image](https://user-images.githubusercontent.com/11585141/106840307-d78a4980-66da-11eb-9f65-670408760dc8.png)

**Comment by jesec** [comment]
I did a brief test on my machine. It turns out that I can successfully download a file (ubuntu-20.04.1-desktop-amd64.iso, 2.59GB) with `rtorrent-flood` image with a hard memory limit of 40MB. Meanwhile, Flood works perfectly fine. 

```
CONTAINER ID   NAME             CPU %     MEM USAGE / LIMIT   MEM %     NET I/O          BLOCK I/O   PIDS
42e9ddcd81b0   rtorrent-flood   17.62%    39.99MiB / 40MiB    99.98%    765MB / 18.4MB   0B / 0B     13
```

So, yeah, there is no need to worry about not-committed, not-allocated memory. It does not reflect the real memory consumption. If you are really paranoid about memory, use Docker's `--memory` argument to hard limit the memory usage of a container.

**Comment by rubenelshof** [comment]
Thank you for the detailed response and explanation.
I indeed got confused with the VSZ and RSS numbers and have read upon it a bit further.
